chips/esp32/src/uart.rs

```diff
@@ -310,14 +306,36 @@ impl<'a> Uart<'a> {
         }
     }
 
+    fn rx_progress(&self) {
+        let regs = self.registers;
+        let idx = self.rx_index.get();
+        let len = self.rx_len.get();
+
+        if idx < len {
+            // Read from the UART hardware and write them to the receive buffer
+            // until either the buffer is full or the UART hardware is empty.
+            self.rx_buffer.map(|rx_buf| {
+                let rx_len = len - idx;
+
+                for i in 0..rx_len {
+                    if regs.status.read(STATUS::RXFIFO_CNT) == 0 {
+                        break;
+                    }
+                    let rx_idx = idx + i;
+                    rx_buf[rx_idx] = regs.fifo.read(FIFO::RXFIFO_RD_BYTE) as u8;
+                    self.rx_index.set(rx_idx + 1)
+                }
+            });
+        }
+    }
+
     pub fn handle_interrupt(&self) {
         let regs = self.registers;
         let intrs = regs.int_st.extract();
 
         if intrs.is_set(INT::TXFIFO_EMPTY_INT) {
-            self.disable_tx_interrupt();
```

Comment:
You aren't clearing the interrupt with this change

Reply:
I did that initially, but @hudson-ayers suggested disabling them only when the transmit or receive is done (lines 338 and 354). I might be missing something though.

Reply:
@alistair23 why would you disable the interrupt if you are still transmitting / still receiving?

Reply:
You don't have to disable the interrupt, but you need to clear the pending interrupt status. Currently this is done in the `disable_*()` functions

Reply:
Oh, I see. I suppose this should be changed back, though renaming the function to `disable_and_clear_interrupts()` might make this clearer.

Reply:
I have split the function in two.